app/views/assets/sharing/batch_sharing_permission_preview.html.erb

```diff
@@ -271,7 +185,121 @@
+                      this.asset_key = change_all(this.asset_key,false)
+                  }.bind(this));
+            };
+            this.teardown = function () {
+                if(this.settings.split_btn_checkbox) {
+                    this.element.find(".jstree-split_btn_checkbox").remove();
+                }
+                parent.teardown.call(this);
+            };
+            this.redraw_node = function(obj, deep, callback, force_draw) {
+                obj = parent.redraw_node.call(this, obj, deep, callback, force_draw);
+                if(obj.classList.contains('asset-node-row')) {
+                    let sbtn_checkbox = document.createElement("span",)
+                    let asset_id=obj.getAttribute("data-node-id")
+                    let sbc_html = ""
+                    if(this.asset_key.find(o => o.data_node_id === asset_id).has_children){
+                        sbc_html = "<%= escape_javascript( render :partial => 'general/split_button_checkbox',
+                                                                    :locals => { checkbox_id: "_asset_id_",
+                                                                                 checkbox_class: "_asset_id_",
+                                                                                 checked: "true" ,
+                                                                                 toggle: "true",
+                                                                                 extra_nesting: 0 } ) %>"
+                    } else {
+                        sbc_html = "<%= escape_javascript( render :partial => 'general/split_button_checkbox',
+                                                                    :locals => { checkbox_id: "_asset_id_",
+                                                                                 checkbox_class: "_asset_id_",
+                                                                                 checked: "true" ,
+                                                                                 toggle: "false",
+                                                                                 extra_nesting: 0 } ) %>"
+                    }
+                    sbc_html = sbc_html.replaceAll("name=\"_asset_id_\"",
+                                    "name=\""+this.share+"["+asset_id.split("-")[0]+"]["+asset_id.split("-")[1]+"]\"")
+                    sbc_html = sbc_html.replaceAll("_asset_id_", asset_id)
+                    if(!(this.asset_key.find(o => o.data_node_id === asset_id).checkbox_status)) {
+                        sbc_html = sbc_html.replaceAll("checked=\"checked\"", "")
+                    }
+                    sbtn_checkbox.innerHTML=sbc_html
+                    let tmp = sbtn_checkbox.cloneNode(true);
+                    obj.insertBefore(tmp, obj.childNodes[1]);
+                }
+                return obj;
+            };
+        };
+    })(jQuery);
+    function get_asset_key(data){
```

Comment:
Correct me if I'm wrong, but from what I can tell, this and the following functions are to preserve checkbox state when nodes are collapsed/expanded (because jstree removes them from the DOM).

I feel like there is a simpler way though. I seem to recall you can store arbitrary `data`  for each node in the jstree. Could you not just store and read the checkbox state on each node rather than in a single structure (that you need to traverse manually).